File: src/negotiate/plan.js
/**
 * Negotiation plan used by the Netlify edge function.
 *
 * 406 is only when the client rejects every type we offer (HTML and
 * Markdown), e.g. Accept: application/pdf. A missing path still has a
 * Markdown representation (/404.md), so Accept: text/markdown must not
 * 406 just because no page-specific sibling exists.
 */

import { MARKDOWN_TYPE, PRODUCES, preferredType } from "./accept.js";

/**
 * @param {{
 *   accept: string | null | undefined,
 *   pageMdExists: boolean,
 *   originStatus?: number,
 * }} input
 * @returns {{
 *   action: "not-acceptable" | "page-markdown" | "not-found-markdown" | "not-found-html" | "origin",
 *   status: number,
 *   chosen: string | null,
 * }}
 */
export function planNegotiation({ accept, pageMdExists, originStatus }) {
  const chosen = preferredType(accept, PRODUCES);
  if (chosen === null) {
    return { action: "not-acceptable", status: 406, chosen: null };
  }
  if (chosen === MARKDOWN_TYPE && pageMdExists) {
    return { action: "page-markdown", status: 200, chosen };
  }
  if (originStatus === 404) {
    if (chosen === MARKDOWN_TYPE) {
      return { action: "not-found-markdown", status: 404, chosen };
    }
    return { action: "not-found-html", status: 404, chosen };
  }
  return { action: "origin", status: originStatus ?? 200, chosen };
}

File: src/melange/plan.ml
(* Negotiation plan used by the Netlify edge function.
   406 is only when the client rejects every type we offer (HTML and
   Markdown). A missing path still has /404.md, so Accept: text/markdown
   must not 406 just because no page-specific sibling exists. *)

type t = { action : string; status : int; chosen : string option }

let plan_negotiation ~accept ~page_md_exists ?origin_status () =
  match Accept.preferred_type accept Accept.produces with
  | None -> { action = "not-acceptable"; status = 406; chosen = None }
  | Some chosen when chosen = Accept.markdown_type && page_md_exists ->
      { action = "page-markdown"; status = 200; chosen = Some chosen }
  | Some chosen when origin_status = Some 404 ->
      if chosen = Accept.markdown_type then
        { action = "not-found-markdown"; status = 404; chosen = Some chosen }
      else { action = "not-found-html"; status = 404; chosen = Some chosen }
  | Some chosen ->
      {
        action = "origin";
        status = (match origin_status with Some s -> s | None -> 200);
        chosen = Some chosen;
      }
